10/README.md

```diff
@@ -1,5 +1,184 @@
+type config struct {
+    // OpenAI API key. You can generate an API key at
+    // https://platform.openai.com/account/api-keys.
+    APIKey string `toml:"api_key"`
+}
+```
+
+Then, embed `weaver.WithConfig[config]` in the struct that implements the
+`ChatGPT` component.
+
+<details>
+<summary>Solution.</summary>
+
+https://github.com/ServiceWeaver/workshops/blob/9beeaf11ec73ad8ab467f9e565ad4adf4e608699/10/chatgpt.go#L31-L42
+</details>
+
+Next, add your secret key to the `api_key` field in the `"emojis/ChatGPT"`
+section of `config.toml`. Again, **be careful not to push your secret key to a
+public repository**.
+
+```toml
+[serviceweaver]
+binary = "./emojis"
+
+["emojis/ChatGPT"]
+api_key = "YOUR SECRET KEY GOES HERE"
+```
+
+Finally, update the implementation of the `Complete` method to read your secret
+key from the config file, returning an error if the key is missing.
+
+<details>
+<summary>Solution.</summary>
+
+https://github.com/ServiceWeaver/workshops/blob/9beeaf11ec73ad8ab467f9e565ad4adf4e608699/10/chatgpt.go#L45-L51
+</details>
+
+## Putting It All Together
+
+In `searcher.go`, add a `SearchChatGPT(context.Context, query string) ([]string,
+error)` method to the `Searcher` component. The `SearchChatGPT` method acts like
+the `Search` method&mdash;it receives a query and returns a list of matching
+emojis&mdash;but it returns the emojis suggested by ChatGPT.
+
+To implement the `SearchChatGPT` method, form a ChatGPT prompt from the user's
+query and pass it to the `ChatGPT.Complete` method. You can get creative with
+the wording of your prompt, but we went with `"Give me a list of emojis that
+related to the query $QUERY. Don't give an explanation."`.
+
+<details>
+<summary>Solution.</summary>
+
+https://github.com/ServiceWeaver/workshops/blob/9beeaf11ec73ad8ab467f9e565ad4adf4e608699/10/searcher.go#L110-L117
+</details>
+
+Next, parse and return the emojis from ChatGPT's response. This is surprisingly
+tricky, as some emojis are composed of multiple unicode code points. We
+recommend using the https://github.com/rivo/uniseg library to segment ChatGPT's
+response into a set of graphemes and checking to see if each grapheme is in the
+`emojis` map. Here's our solution:
+
+https://github.com/ServiceWeaver/workshops/blob/9beeaf11ec73ad8ab467f9e565ad4adf4e608699/10/searcher.go#L128-L138
+
+In `main.go`, add a `/search_chatgpt?q=<query>` endpoint that calls the
+`SearchChatGPT` method.
+
+<details>
+<summary>Solution.</summary>
+
+https://github.com/ServiceWeaver/workshops/blob/9beeaf11ec73ad8ab467f9e565ad4adf4e608699/10/main.go#L61-L63
+</details>
+
+Finally, build and run your application:
+
+```
+$ weaver generate .
+$ weaver multi deploy config.toml
+```
+
+You can curl the `/search_chatgpt` endpoint directly:
+
+```console
+$ curl localhost:9000/search_chatgpt?q=happy
+["😀","😁","😃","😄","😊","😍","😎","🤗","😻","🌞","🎉","🎊","🎁","🎈","💐","👍","✨","🌟","💫","🌈"]
+```
+
+Though, we strongly recommend you use the web UI provided in [Part 4](../04).
```

Comment:
Just checking if this UI was introduced in Part 5?

Reply:
Should be Part 4.